### PR TITLE
Automatically generate driver crates

### PR DESCRIPTION
**Pull request by jessebraham**
I saw #33 this morning and thought it sounded fun, so I threw this together real quick 😉 

I ended up using `tinytemplate` and `cargo-task` as you suggested. You should be able to try it out by running:
```bash
$ cargo install cargo-task
$ cargo task generate-drivers
```
I do make the assumption that this is being executed from the Step/Dir project's root, just FYI.

It's pretty basic and could probably be improved, but it gets the job done for now. I tried to keep things simple and not over-engineer it. There's some code duplication, but I figured it was probably okay because I can't imagine it will change drastically.

I'm open to any feedback you may have, I rarely have access to `std` when I write Rust so I may be doing some silly things. If there's anything you don't like or anything you'd like added, just let me know and I'm happy to change anything.

### CODE REVIEW

File: drivers.toml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# If adding a new driver, please try to keep the entries alphabetized! :)
+
+[drivers.drv8825]
+name = "drv8825"
```
Comment (hannobraun) [inline drivers.toml:4]: Isn't that redundant? You get the driver name from the map key, as well as the struct. Probably best to choose the one that is most useful in the code and remove the other.

So either:
``` toml
[drivers.drv8825]
# no `name`
# ...
```

Or:
``` toml
[[drivers]]
name = "drv8825"
# ...
```

File: drivers.toml [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+version = "0.3.0"
+authors = [
+  "Hanno Braun <hanno@braun-embedded.com>",
+  "Jesse Braham <jesse@beta7.io>",
+]
```
Comment (hannobraun) [inline drivers.toml:9]: The ideal implementation would get all of this information from the `step-dir` Cargo.toml. I realize that not all the drivers have the same `authors` currently, but I don't think it's worth the effort to track them separately (anything of substance lives in `step-dir` proper).

File: templates/driver/README.md.tmpl [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# { name | upper } Driver [![crates.io](https://img.shields.io/crates/v/{ name }.svg)](https://crates.io/crates/{ name }) [![Documentation](https://docs.rs/{ name }/badge.svg)](https://docs.rs/{ name }) ![CI Build](https://github.com/braun-embedded/step-dir/workflows/CI%20Build/badge.svg)
+
+## About
+
+Rust driver crate for the [{ name | upper }] stepper motor driver. Carrier boards for this chip are [available from Pololu].
+
+This crate is a specialized facade of the [Step/Dir] library. Please consider using Step/Dir directly, as it provides drivers for more stepper motor drivers, as well as an interface to abstract over them.
```
Comment (hannobraun) [inline templates/driver/README.md.tmpl:7]: Not really on topic here (I suppose that's the same text as before), but this sounds better to me:
```suggestion
This crate is a specialized facade for the [Step/Dir] library. Please consider using Step/Dir directly, as it provides drivers for more stepper motor drivers, as well as an interface to abstract over them.
```

Feel free to disregard, I'm not the native speaker here :-)

File: templates/driver/README.md.tmpl [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+See [Step/Dir] for more documentation and usage examples.
+
+## Status
+
+This driver is currently very basic in its capabilities. Its design is experimental, and more revisions to the API are expected.
```
Comment (hannobraun) [inline templates/driver/README.md.tmpl:13]: Again, not completely on topic, but might be best to remove this. We already link to Step/Dir, so it's probably best to keep the information we need to maintain here to a reasonable minimum.

File: templates/driver/src/lib.rs.tmpl [hunk 5]
```diff
@@ -0,0 +1,15 @@
+//! { name | upper } Driver
+//!
+//! Platform-agnostic driver library for the { name | upper } stepper motor driver.
+//! This crate is a specialized facade of the [Step/Dir] library. Please
```
Comment (hannobraun) [inline templates/driver/src/lib.rs.tmpl:4]:
```suggestion
//! This crate is a specialized facade for the [Step/Dir] library. Please
```
Again, not on topic, feel free to disregard.

File: .cargo-task/generate-drivers/src/config.rs [hunk 6]
```diff
@@ -0,0 +1,39 @@
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{prelude::*, Result};
+use std::path::PathBuf;
+
+use serde_derive::Deserialize;
+
+// Values required to populate the 'Cargo.toml', 'lib.rs', and 'README.md'
+// templates for a given driver implementation.
```
Comment (hannobraun) [inline .cargo-task/generate-drivers/src/config.rs:9]:
```suggestion
/// Values required to populate the 'Cargo.toml', 'lib.rs', and 'README.md'
/// templates for a given driver implementation.
```

File: .cargo-task/generate-drivers/src/config.rs [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+#[derive(Debug, Deserialize)]
+pub struct Driver {
+    pub name: String,
+    pub version: String,
+    pub authors: Vec<String>,
+    pub product_url: String,
+    pub pololu_url: String,
+}
+
+// The 'drivers.toml' file format.
```
Comment (hannobraun) [inline .cargo-task/generate-drivers/src/config.rs:19]:
```suggestion
/// The 'drivers.toml' file format.
```

File: .cargo-task/generate-drivers/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+mod cargo_task_util;
+
+mod config;
+use config::{load_drivers_toml, Driver};
+
+fn main() -> std::io::Result<()> {
+    // `root`      - executing directory; assumed to be Step/Dir root
+    // `drivers`   - driver facade crate directory (output)
+    // `templates` - template directory (input)
+    let root = env::current_dir()?;
```
Comment (hannobraun) [inline .cargo-task/generate-drivers/src/main.rs:21]: Idea for future extension: Scan directory tree, looking for `drivers.toml`. Use that directory for `root`.

**Review by hannobraun — APPROVED** [review]
Wow, this is awesome! I figured it would make sense to share the more well-thought out parts of my plans publicly, but I certainly didn't expect to wake up to a completed task the next day! :grin: 

I left some comments. **All of this is nit-picking.** I'd be more than happy to merge this as-is. I'm leaving it unmerged, in case you want to implement any of my suggestions right now. If not, feel free to just merge as soon as you read this.

These are some of the things I suggested in the comments. Perfectly fine to not do these right now, but if this gets merged without them, I'd like to open issues to track them, so here's a list:

- [ ] redundant driver names
- [ ] read `version` and `author` from main `Cargo.toml`
- [ ] automatically find root directory by scanning for `drivers.toml`

And here are some additional suggestions I'd also like to track, in case the pull request is merged without them:

- [ ] Might be best to leave the `Cargo.lock` out of the repository. This will cost us some stability, but this code isn't relevant to end users anyway, and not having `Cargo.lock` checked in might flush out potential problems due to dependency updates more quickly.
- [ ] This should be added to `scripts/build.sh`.
- [ ] Should be documented in `CONTRIBUTING.md`.
- [ ] Would be nice to have a `README.md` in `drivers/` that tells people not to make changes and refers to the documentation in `CONTRIBUTING.md`.

As for general feedback on the `std`-based code: Looks good to me. One thing that stands out is the duplication between `Driver` and `Context` (which is assume is what you were referring to), but I don't know how to improve it, so whatever.

One thing I'd have done differently are the `expect` calls. Perfectly fine to panic on errors, in this kind of code, but handling it via `?` reads nicer. Again, just nit-picking since you asked for feedback. This is perfectly fine as-is.

**Comment by jessebraham** [comment]
Thank you for the feedback! I think I have fixed most of the issues you raised.

I will leave the searching for `drivers.toml` and parsing of Step/Dir's `Cargo.toml` to another PR if that's okay. If you'd like I can add documentation here, otherwise it can be added at another time as well.

**Comment by hannobraun** [comment]
Thanks again, the changes look great!


> I will leave the searching for `drivers.toml` and parsing of Step/Dir's `Cargo.toml` to another PR if that's okay. If you'd like I can add documentation here, otherwise I can add it at another time as well.

Of course! I'll open issues to track further improvements after merging this.

**Pull request merged**